File: other/MT1051/week03/warm_up/find_root.py
# -*- coding: utf-8 -*-
"""
Date: 2023/09/25
Warm up
Author: 林群賀
Student Number: 109601003
"""

import math
from typing import Any

def find_roots(a, b, c):
    discriminant = b ** 2 - 4 * a * c
    
    if discriminant > 0:
        root1 = (-b + math.sqrt(discriminant)) / (2 * a)
        root2 = (-b - math.sqrt(discriminant)) / (2 * a)
        return (
            round(root1, 2), 
            round(root2, 2)
        )
    elif discriminant == 0:
        root1 = -b / (2 * a)
        return round(root1, 2)
    else:
        return "No real roots"

if __name__ == "__main__":
    try:
        a = float(input("Enter the value of a: "))
        b = float(input("Enter the value of b: "))
        c = float(input("Enter the value of c: "))
        
        roots = find_roots(a, b, c)
        
        if isinstance(roots, tuple):
            print(f"The roots of the equation are {roots[0]} and {roots[1]}")
        elif isinstance(roots, float):
            print(f"The root of the equation is {roots}")
        else:
            print(roots)
    except ValueError:
        print("Invalid input. Please enter valid numerical values for a, b, and c.")

class FindRoot:
    def __call__(self, *args: Any, **kwds: Any) -> Any:
        pass

    def __init__(self) -> None:
        pass

    def get_parameters(self,) -> tuple:
        a = float(input("Enter the value of a: "))
        b = float(input("Enter the value of b: "))
        c = float(input("Enter the value of c: "))

        return a, b, c
    
    def get_root(self, a, b, c):
        discriminant = b ** 2 - 4 * a * c
    
        if discriminant > 0:
            root1 = (-b + math.sqrt(discriminant)) / (2 * a)
            root2 = (-b - math.sqrt(discriminant)) / (2 * a)
            return (
                round(root1, 2), 
                round(root2, 2)
            )
        elif discriminant == 0:
            root1 = -b / (2 * a)
            return round(root1, 2)
        else:
            return "No real roots"

File: mt/mt1051/week03/warm_up/find_root.go
// Date: 2023/09/25
// Warm up
// Author: 林群賀
// Student Number: 109601003

package main

import (
	"fmt"
	"math"
)

func findRoots(a, b, c float64) interface{} {
	discriminant := b*b - 4*a*c

	if discriminant > 0 {
		root1 := (-b + math.Sqrt(discriminant)) / (2 * a)
		root2 := (-b - math.Sqrt(discriminant)) / (2 * a)
		return []float64{math.Round(root1*100) / 100, math.Round(root2*100) / 100}
	} else if discriminant == 0 {
		root1 := -b / (2 * a)
		return math.Round(root1*100) / 100
	} else {
		return "No real roots"
	}
}

func main() {
	var a, b, c float64
	fmt.Print("Enter the value of a: ")
	fmt.Scan(&a)
	fmt.Print("Enter the value of b: ")
	fmt.Scan(&b)
	fmt.Print("Enter the value of c: ")
	fmt.Scan(&c)

	roots := findRoots(a, b, c)

	switch r := roots.(type) {
	case []float64:
		fmt.Printf("The roots of the equation are %.2f and %.2f\n", r[0], r[1])
	case float64:
		fmt.Printf("The root of the equation is %.2f\n", r)
	case string:
		fmt.Println(r)
	}
}
